indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java

```diff
@@ -0,0 +1,566 @@
+                              synchronized (tasks) {
+                                taskWorkItem = tasks.get(task.getId());
+
+                                if (taskWorkItem == null) {
+                                  LOGGER.makeAlert("WTF?! TaskInfo disappeared!").addData("task", task.getId()).emit();
+                                  throw new ISE("TaskInfo disappeared for task[%s]!", task.getId());
+                                }
+
+                                if (taskWorkItem.shutdown) {
+                                  throw new IllegalStateException("Task has been shut down!");
+                                }
+                              }
+
+                              if (!taskFile.exists()) {
+                                jsonMapper.writeValue(taskFile, task);
+                              }
+
+                              // This will block for a while. So we append the thread information with more details
+                              final String priorThreadName = Thread.currentThread().getName();
+                              Thread.currentThread().setName(StringUtils.format("%s-[%s]", priorThreadName, task.getId()));
+
+                              TaskStatus taskStatus = null;
+                              final TaskToolbox toolbox = toolboxFactory.build(task);
+                              try {
+                                ListenableFuture<TaskStatus> taskStatusFuture = taskExecutor.submit(
+                                    new Callable<TaskStatus>()
+                                    {
+                                      @Override
+                                      public TaskStatus call()
+                                      {
+                                        taskWorkItem.setThread(Thread.currentThread());
+                                        try {
+                                          return task.run(toolbox);
+                                        }
+                                        catch (Exception e) {
+                                          LOGGER.error(e, "Task[%s] exited with exception.", task.getId());
+                                          return null;
+                                        }
+                                        finally {
+                                          taskWorkItem.setThread(null);
+                                        }
+                                      }
+                                    }
+                                );
+                                TaskRunnerUtils.notifyLocationChanged(listeners, task.getId(), taskLocation);
+                                TaskRunnerUtils.notifyStatusChanged(
+                                    listeners,
+                                    task.getId(),
+                                    TaskStatus.running(task.getId())
+                                );
+                                taskStatus = taskStatusFuture.get();
```

Comment:
not sure why we are running task in a separate thread while this thread is waiting for the task to finish anyway ?

if this is done for making calls to `notifyXXX()` then they could be done right before calling `task.run(toolbox)` if task was run in this thread itself ?

Reply:
It was partly for making the notify calls after the run started, and partly because I wanted to have a similar structure to ForkingTaskRunner where the control context and the task code were separated by thread/process, but it's not really necessary to do that.

I changed the task setup and run() to execute using a thread in `taskExecutor`, the control thread pool is only used to run externally triggered task shutdown now (and there is only 1 per task slot now)